File: POO/Constructores_Clases/Abstract class/JAVA/Figura.java
public abstract class Figura {
    protected String color;

    public Figura() {
        this.color = "negro";
    }

    public abstract double calcularArea();
    public abstract double calcularPerimetro();

    public void mostrar() {
        System.out.println("Color: " + color);
    }
}

class Rectangulo extends Figura {
    private double ancho;
    private double alto;

    public Rectangulo() {
        super();
        this.ancho = 1.0;
        this.alto = 1.0;
    }

    @Override
    public double calcularArea() {
        return ancho * alto;
    }

    @Override
    public double calcularPerimetro() {
        return 2 * (ancho + alto);
    }

    @Override
    public void mostrar() {
        super.mostrar();
        System.out.println("Rectángulo - Ancho: " + ancho + ", Alto: " + alto);
        System.out.printf("Área: %.2f%n", calcularArea());
        System.out.printf("Perímetro: %.2f%n", calcularPerimetro());
    }

    public static void main(String[] args) {
        Figura rectangulo = new Rectangulo();
        rectangulo.mostrar();
    }
}

File: POO/Constructores_Clases/Abstract class/js/Figura.js
class Figura {
    constructor() {
        if (this.constructor === Figura) {
            throw new Error("No se puede instanciar una clase abstracta");
        }
        this.color = "negro";
    }

    calcularArea() {
        throw new Error("Método abstracto debe ser implementado");
    }

    calcularPerimetro() {
        throw new Error("Método abstracto debe ser implementado");
    }

    mostrar() {
        console.log(`Color: ${this.color}`);
    }
}

class Circulo extends Figura {
    constructor() {
        super();
        this.radio = 1.0;
    }

    calcularArea() {
        return Math.PI * this.radio * this.radio;
    }

    calcularPerimetro() {
        return 2 * Math.PI * this.radio;
    }

    mostrar() {
        super.mostrar();
        console.log(`Círculo - Radio: ${this.radio}`);
        console.log(`Área: ${this.calcularArea().toFixed(2)}`);
        console.log(`Perímetro: ${this.calcularPerimetro().toFixed(2)}`);
    }
}

// Uso
const circulo = new Circulo();
circulo.mostrar();